icicle/backend/cpu/include/ntt_tasks.h

```diff
@@ -0,0 +1,1260 @@
+        j = (ntt_task_cordinates.hierarchy_0_layer_idx == 0) ? ntt_task_cordinates.hierarchy_0_subntt_idx
+                                                             : ntt_task_cordinates.hierarchy_0_block_idx;
+        uint64_t tw_idx = (this->direction == NTTDir::kForward)
+                            ? ((this->domain_max_size / ntt_size) * j * i)
+                            : this->domain_max_size - ((this->domain_max_size / ntt_size) * j * i);
+        elements_of_current_batch[elem_mem_idx] = elements_of_current_batch[elem_mem_idx] * this->twiddles[tw_idx];
+      }
+    }
+  }
+
+  /**
+   * @brief Pushes tasks for the h1 hierarchy of the NTT computation.
+   *
+   * This function organizes and pushes tasks for the h1 hierarchy level of the NTT computation
+   * into the task manager. It iterates over the layers and sub-NTTs, scheduling the necessary
+   * computations while ensuring that the data elements remain close in memory for efficiency,
+   * assuming that in this hierarchy the sub-NTT fits in the cache.
+   * If multiple layers are involved, Instead of sorting between layers, we are using the function
+   * idx_in_mem to calculate the memory index of each element a final reorder task is pushed to fix
+   * the order after processing.
+   *
+   * @param input The input array of elements to be processed.
+   * @param ntt_task_cordinates The coordinates specifying the current sub-NTT within the NTT hierarchy.
+   * @param ntt_tasks_manager The task manager responsible for handling the scheduling of tasks.
+   * @return eIcicleError Returns `SUCCESS` if all tasks are successfully pushed to the task manager.
+   */
+  template <typename S, typename E>
+  eIcicleError NttCpu<S, E>::hierarchy1_push_tasks(
+    E* input, NttTaskCordinates ntt_task_cordinates, NttTasksManager<S, E>& ntt_tasks_manager)
+  {
+    uint64_t original_size = (this->ntt_sub_logn.size);
+    int nof_h0_layers =
+      (NttCpu<S, E>::ntt_sub_logn.hierarchy_0_layers_sub_logn[ntt_task_cordinates.hierarchy_1_layer_idx][2] != 0)   ? 3
+      : (NttCpu<S, E>::ntt_sub_logn.hierarchy_0_layers_sub_logn[ntt_task_cordinates.hierarchy_1_layer_idx][1] != 0) ? 2
+                                                                                                                    : 1;
+    for (ntt_task_cordinates.hierarchy_0_layer_idx = 0;
+         ntt_task_cordinates.hierarchy_0_layer_idx <
+         NttCpu<S, E>::ntt_sub_logn.hierarchy_0_layers_sub_logn[ntt_task_cordinates.hierarchy_1_layer_idx].size();
+         ntt_task_cordinates.hierarchy_0_layer_idx++) {
+      if (ntt_task_cordinates.hierarchy_0_layer_idx == 0) {
+        int log_nof_subntts =
+          NttCpu<S, E>::ntt_sub_logn.hierarchy_0_layers_sub_logn[ntt_task_cordinates.hierarchy_1_layer_idx][1];
+        int log_nof_blocks =
+          NttCpu<S, E>::ntt_sub_logn.hierarchy_0_layers_sub_logn[ntt_task_cordinates.hierarchy_1_layer_idx][2];
+        for (ntt_task_cordinates.hierarchy_0_block_idx = 0;
```

Comment:
double loop can be simplified into a single loop using mod (shifts) and then all this code can become a single loop that runs on layer idx